README.md
[![Code Health](https://landscape.io/github/yafp/gkanapass/master/landscape.svg?style=flat)](https://landscape.io/github/yafp/gkanapass/master)
[![License](https://img.shields.io/badge/license-GPL3-brightgreen.svg)](LICENSE)

![logo](https://raw.githubusercontent.com/yafp/gkanapass/master/doc/fa-unlock-alt_64_0_000000_none.png) gkanapass
==========

![xkcd](https://raw.githubusercontent.com/yafp/gkanapass/master/doc/xkcd_936_password_strength.png)
https://xkcd.com/936/

## About
gkanapass is a python based password generator influenced by [kana](https://en.wikipedia.org/wiki/Kana)


You might ask
> why another password generator?

I was in need for a python playground, simple as that.

## Install / Uninstall
- Unpack the archive
- navigate to folder which contains this README.md
- Install by running:

> sudo make install

Uninstall by running:
> sudo make uninstall


## Usage
To display the general help:
> gkanapass -h

To generate passwords with default length (10)
> gkanapass

To generate passwords with user-defined length (14)
> gkanapass 14

![Usage](https://raw.githubusercontent.com/yafp/gkanapass/master/doc/gkanapass_usage_optimized.gif)

__Be aware:__  gkanapass is forcing a min password length of 8.


## Supported platforms
gkanapass was tested so far on the following setups/platforms
- Linux (using python2 and python3)


CHANGELOG.md
![logo](https://raw.githubusercontent.com/yafp/gkanapass/master/doc/fa-unlock-alt_64_0_000000_none.png) gkanapass
==========

# CHANGELOG
## 2017
### 20171221.01
- Fixed typo in help (#12)
- Adding leading zero to password numbers
- Password number starts with 1 now
- Password generation modified (results are less human readable)

## 2016
### 20160929.05
- Adding several new chars to special char pool

### 20160929.04
- Fixing issues with man and makefile

### 20160929.03
- Fixed minor issue in makefile regarding manpage

### 20160929.02
- Fixing missing parentheses in print

### 20160929.01
- Fixing #11 (display password length informations in output)

### 20160408.01
- Adding project url to help dialog & several minor changes to password generation.

### 20160309.02
- Adding app icon to readme - fixing #10

### 20160309.01
- Adding an project icon file

### 20160222.01 
- Fixing #7 (adding a man page) 

### 20160220.02 
- fixing usage gif filename - which is referenced in readme

### 20160220.01 
- Adding a license sticker to project readme

### before
- ...



src/functions.py
#!/usr/bin/env python2
# coding=utf-8
#
# Name:			functions.py
# Function:		Defines the functions for gkanapass.py
# Date:			20160210
# Author: 		yafp

import sys		# for handling arguments
import random


# --------------------------------------------------------------------
# validate the arguments
# --------------------------------------------------------------------
def validateArguments():
	#print 'Number of arguments:', len(sys.argv), 'arguments.'
	#print 'Argument List:', str(sys.argv)

	if len(sys.argv) < 2: # no user defined length available
		print("- Setting password length to default (10)")
		length="10"
		#sys.exit('Usage: %s database-name' % sys.argv[0])

	else: # user entered an argument - check it
		length=sys.argv[1]
		print("- Setting password length (" + length +")")

		if int(float(length))<8: # check if argument is smaller 8
			print("- Forcing min password length (8)")
			length="8"

	return length;



# --------------------------------------------------------------------
# generate a password
# --------------------------------------------------------------------
def generateKanaPass(length):

	print "\nPasswords:"
	# define possible symbols
	#charsPureKana = ['k','s','t','n','h','m','y','r','w']
	charsPureKana = ['k','s','t','n','h','m','y','r','w','K','S','T','N','H','M','Y','R','W']
	#charsNonPureKana = ['k','s','t','n','h','m','y','r','w','K','S','T','N','H','M','Y','R','W','d','f','l','p','v','D','F','L','P','V']

	#charsPureVocal = ['a','e','i','o','u']
	charsPureVocal = ['a','e','i','o','u','A','E','I','O','U']


	#charsVocalNumbers = ['a','e','i','o','u','A','E','I','O','U','1','2','3','4','5','6','7','8','9','0']

	for i in range (1,11): # generate 10 passwords
		generatedPassword=''
		for num in range (0,length): # build single password
			generatedPassword=generatedPassword+random.choice(charsPureKana)+random.choice(charsPureVocal)


		# hack: user-defined length might be odd -> generated password is always even -> cut to defined length
		#intLength=int(float(length))
		generatedPassword=generatedPassword[:int(float(length))]
		print str(i) + ":\t" + generatedPassword


src/gkanapass.py
#!/usr/bin/env python2
# coding=utf-8
#
# Name:			gkanapass.py
# Function:		a python based password generator with focus on japanese/kana-style passwords
#               https://en.wikipedia.org/wiki/Kana
#               https://xkcd.com/936/
# Date:			20171221.01
# Author: 		yafp


# todo
# - add try & except blocks
# - harden parameter validation


#==========================     IMPORT     =====================================
import random		# for randomness
import os			# for clearing the screen
import sys			# for handling arguments
#import argparse	# better argument parsing
					# https://stackoverflow.com/questions/4188467/how-to-check-if-an-argument-from-commandline-has-been-set


#=========================     VARIABLES     ===================================
appName="gkanapass"
appDescription="gkanapass is a python based password generator influenced by kana"
appDevURL="https://github.com/yafp/gkanapass"
appVersion="20171221.01"


#=========================     FUNCTIONS     ===================================
def clearScreen():
	if (os.name == 'posix') or (os.name == 'mac'): # linux or mac
		os.system('clear')
	if os.name == 'nt': # windows
		os.system('cls')


def printHead(title):
	#clearScreen()
	print ("------------------------------------------------")
	print (" \033[1m"+ appName + "\033[0m - (" + appVersion +") - " + title)
	print ("------------------------------------------------\n")
	return;


def validateArguments():
	#print 'Number of arguments:', len(sys.argv), 'arguments.'
	#print 'Argument List:', str(sys.argv)

	pwLengthInfo="unset"

	if len(sys.argv) < 2: # no user argument available
		pwLengthInfo="Password length: 10 (default)"
		length="10"				# default length

	elif (sys.argv[1]=="-v"):	# version
		displayVersion()	# display version informations
		sys.exit()

	elif (sys.argv[1]=="-h"):	# help 
		displayHelp()	# display help informations
		sys.exit()

	else: # user entered an argument - check it
		length=sys.argv[1]
		try:
			if int(float(length))>7: # if argument is bigger then 7 -> good
				pwLengthInfo="Password length: "+length
			
			if int(float(length))<8: # check if argument is smaller 8 -> bad
				length="8"		# force to min length
				pwLengthInfo="Password length 8 (forced minimal value)"
		except BaseException:	# invalid parameter (not a number) - jump back to default
			pwLengthInfo="Password length: 10 (forcing default value)"
			length="10"			# default length

	print (pwLengthInfo+"\n")
	return length;


def generateKanaPass(length):

	# define some symbol-pools
	charPoolConsonantKana = ['k','s','t','n','h','m','y','r','w']		# kana consonants
	charPoolConsonantSimilar = ['d','f','l','p','v']					# non-kana consonants
	charPoolVocal = ['a','e','i','o','u']								# vocals
	charPoolNumbers = ['1','2','3','4','5','6','7','8','9','0']			# numbers
	charPoolSpecials = ['#','+','-','_','.','?','!',"&",";","@","<",">","~","%","$","|"]				# specials

	for i in range (01,11): # generate 10 passwords
		generatedPassword='' # start with an empty password

		for dummy in range (0,length): # build single_generateRandomPair
			randomNumber=random.randrange(1,100) # generate a random number between 1 and 100 - based on that we randomize the password-generation

			if randomNumber>90: 		# lowercase consonants kana + special
				generatedPassword=generatedPassword+random.choice(charPoolConsonantKana)+random.choice(charPoolSpecials)

			elif randomNumber>85:		# uppercase consonants similar & uppercase vocal
				generatedPassword=generatedPassword+random.choice(charPoolConsonantSimilar).upper()+random.choice(charPoolVocal).upper()

			elif randomNumber>75: 		# uppercase consonants kana + number
				generatedPassword=generatedPassword+random.choice(charPoolConsonantKana).upper()+random.choice(charPoolNumbers)

			elif randomNumber>65: 		# lowercase consonants kana + number
				generatedPassword=generatedPassword+random.choice(charPoolConsonantKana)+random.choice(charPoolNumbers)

			elif randomNumber>55:		# uppercase consonants kana & lowercase vocal
				generatedPassword=generatedPassword+random.choice(charPoolConsonantSimilar).upper()+random.choice(charPoolVocal)

			elif randomNumber>45:		# lowercase consonants similar & lowercase vocal
				generatedPassword=generatedPassword+random.choice(charPoolConsonantSimilar)+random.choice(charPoolVocal)

			elif randomNumber>35:		# lowercase kana similar & uppercase vocal
				generatedPassword=generatedPassword+random.choice(charPoolConsonantSimilar)+random.choice(charPoolVocal).upper()

			else: 						# lowercase consonants kana & lowercase vocal
				generatedPassword=generatedPassword+random.choice(charPoolConsonantKana)+random.choice(charPoolVocal)

		# user-defined password length might be odd
		# generated password is always even and too long -> cut generated to defined length
		generatedPassword=generatedPassword[:int(float(length))]

		# output the generated and cutted password
		print ("  "+str(i).zfill(2) + ": " + generatedPassword) # zfill for leading zero
	print("\n") # final newline after password list


def displayHelp():
	clearScreen()
	printHead("Help")
	print ("ABOUT:")
	print ("\t"+appDescription)
	print ("\t"+appDevURL+"\n")
	print ("CORE:")
	print ("\tgkanapass\tGenerates 10 passwords with default length (10)")
	print ("\tgkanapass 14\tGenerates 10 passwords with user-defined length (14)")
	print ("\nMISC:")
	print ("\tgkanapass -h\tDisplay this help dialog")
	print ("\tgkanapass -v\tDisplay application version")


def displayVersion():
	clearScreen()
	printHead("Version")
	print ("VERSION:")
	print ("\t" + appVersion)


#========================     MAIN SCRIPT     ==================================
clearScreen() # clear the screen
printHead("Passwords") # output a head containing appname, version etc
intLength=int(float(validateArguments())) # validate argument & parse to int
generateKanaPass(intLength)	# Generate 10 passwords with defined length
#==========================     THE END     ====================================
